deployment_logger.go

```diff
@@ -0,0 +1,248 @@
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+package main
+
+import (
+	"bufio"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"sort"
+	"strconv"
+)
+
+// error messages
+var (
+	ErrNoLogFiles           = errors.New("log directory empty")
+	ErrLoggerNotInitialized = errors.New("logger not initialized")
+	ErrNoDeploymentId       = errors.New("no valid deployment id")
+	ErrParsingLogFile       = errors.New("can not parse deployment log file")
+)
+
+// JSONFileLogger struct
+type FileLogger struct {
+	logFileName string
+	logFile     io.WriteCloser
+}
+
+// NewJSONFileLogger creates instance of file logger; it is initialized
+// just before logging is started
+func NewFileLogger(name string) *FileLogger {
+	// open log file
+	logFile, err := os.OpenFile(name, os.O_RDWR|os.O_CREATE|os.O_APPEND, 0666)
+	if err != nil {
+		// if we can not open file for logging; return nil
+		return nil
+	}
+
+	// return FileLogger only when logging is possible (we can open log file)
+	return &FileLogger{
+		logFileName: name,
+		logFile:     logFile,
+	}
+}
+
+func (fl *FileLogger) Write(log []byte) (int, error) {
+	return fl.logFile.Write(log)
+}
+
+func (fl *FileLogger) Deinit() error {
+	return fl.logFile.Close()
+}
+
+type DeploymentLogManager struct {
+	logLocation         string
+	logFileName         string
+	logDeploymentIdFile string
+	deploymentStore     *DirStore
+
+	deploymentID string
+	logger       *FileLogger
+	// how many log files we are keeping in log directory before rotating
+	maxLogFiles int
+	// it is easy to add logging hook, but not so much remove it;
+	// we need a mechanism for emabling and disabling logging
+	loggingEnabled bool
+}
+
+func NewDeploymentLogManager(logDirLocation string) *DeploymentLogManager {
+	return &DeploymentLogManager{
+		logLocation: filepath.Dir(logDirLocation),
+		logFileName: filepath.Base(logDirLocation),
+		// for now we can easily hardcode file name
+		logDeploymentIdFile: "last_deployment_id",
+		deploymentStore:     NewDirStore(filepath.Dir(logDirLocation)),
+		// file logger needs to be instanciated just before writing logs
+		//logger:
+		// for now we can hardcode this
+		maxLogFiles:    5,
+		loggingEnabled: false,
+	}
+}
+
+func (dlm DeploymentLogManager) WriteLog(log []byte) error {
+	if dlm.logger == nil {
+		return ErrLoggerNotInitialized
+	}
+	_, err := dlm.logger.Write(log)
+	return err
+}
+
+func (dlm *DeploymentLogManager) Enable() error {
+	if dlm.loggingEnabled {
+		// TODO: maybe in fact we should return error
+		return nil
+	}
+
+	// check if we have valid deploymend id
+	if dlm.deploymentID == "" {
```

Comment:
From the other hand we will need to call `Enable()` at least from two places: 
- when we are starting new update (here we will know what id is as we will receive it from server)
- when we are committing new update (at this point we will have to read it from somewhere; most likely some persistent location as at this point we will be after reboot)
- after some error happened (like we started update; then there was power-loss; and after reboot we need to report back to server that update was broken because ...)

In all cases, but first one, we will need to get deployment id from somewhere.